src/main/java/com/google/api/codegen/config/ResourceDescriptorConfig.java

```diff
@@ -272,14 +272,16 @@ private String getUnqualifiedTypeName() {
   static String getParentPattern(String pattern) {
     List<String> segments = getSegments(pattern);
     int index = segments.size() - 1;
+    if (index < 0) {
+      return "";
+    }
     if (isVariableBinding(segments.get(index))) {
       index -= 2;
     } else {
       index--;
     }
-    if (index < 0) {
-      return "";
-    }
+    Preconditions.checkArgument(
+        index >= -1, "malformatted pattern, can't calculate parent: %s", pattern);
```

Comment:
I suspect this should be `index >= 0` - as otherwise we get an *empty* pattern, which still isn't a useful parent.
(Basically, taking the root of "foo" for "foos/{foo}" is invalid, I'd say. The empty pattern can't represent a resource name, so it's not useful.)

Reply:
That is correct from the public-facing point of view - but this is an internal method. We need to call this method on every resource pattern to find if it has a parent, if so, which resource does the parent belong to. Therefore we can't error if the calculated parent is empty (for instance "projects/{project}". I guess we are sort of using an empty string as a way to express "this pattern does not have a parent pattern".

Reply:
Ah, okay - in C# I believe we only find parents when we're asked to :)

Reply:
Hi Jon, can you approve the PR if you don't think there are other issues?

Reply:
There's still the matter of the parent - of `"*"` - which wasn't clear in my previous comment due to Markdown, unfortunately. Basically I'd like to see a test showing that if you pass in `"*"`, you get `"*"` back - and code to make that work.

Reply:
Oh that solves all the puzzle now. I will fix `"*"` shortly.